macros/graph/plotly3D.pl

```diff
@@ -0,0 +1,626 @@
+			($self->{vMin}, $self->{vMax}, $self->{vStep}) = $self->cmpBounds(shift, 20);
+		} else {
+			($self->{tMin}, $self->{tMax}, $self->{tStep}) = $self->cmpBounds(shift, 100);
+		}
+	}
+}
+
+sub genPoints {
+	my $self  = shift;
+	my $id    = shift || 1;
+	my $count = shift || 1;
+	my $type  = $self->{funcType};
+
+	if ($type eq 'data') {
+		# Manual data plot, nothing to do.
+	} elsif ($type eq 'jsmd' || $type eq 'js') {
+		if ($type eq 'jsmd') {
+			foreach ('xFunc', 'yFunc', 'zFunc') {
+				$self->{$_} = $self->funcToJS($self->{$_});
+			}
+		}
+		$self->{xPoints} = "xData${id}_$count";
+		$self->{yPoints} = "yData${id}_$count";
+		$self->{zPoints} = "zData${id}_$count";
+	} elsif ($type eq 'perl') {
+		$self->buidArray;
+	} else {
+		Value::Error("Unkown plot type: $type\n");
+	}
+}
+
+# Takes a pseudo function string and replaces with JavaScript functions.
+sub funcToJS {
+	my $self   = shift;
+	my $func   = shift;
+	my %vars   = map { $_ => $_ } @{ $self->{variables} };
+	my %tokens = (
+		sqrt    => 'Math.sqrt',
+		cbrt    => 'Math.cbrt',
+		hypot   => 'Math.hypot',
+		norm    => 'Math.hypot',
+		pow     => 'Math.pow',
+		exp     => 'Math.exp',
+		abs     => 'Math.abs',
+		round   => 'Math.round',
+		floor   => 'Math.floor',
+		ceil    => 'Math.ceil',
+		sign    => 'Math.sign',
+		int     => 'Math.trunc',
+		log     => 'Math.ln',
+		ln      => 'Math.ln',
+		cos     => 'Math.cos',
+		sin     => 'Math.sin',
+		tan     => 'Math.tan',
+		acos    => 'Math.acos',
+		arccos  => 'Math.acos',
+		asin    => 'Math.asin',
+		arcsin  => 'Math.asin',
+		atan    => 'Math.atan',
+		arctan  => 'Math.atan',
+		atan2   => 'Math.atan2',
+		cosh    => 'Math.cosh',
+		sinh    => 'Math.sinh',
+		tanh    => 'Math.tanh',
+		acosh   => 'Math.acosh',
+		arccosh => 'Math.arccosh',
+		asinh   => 'Math.asinh',
+		arcsinh => 'Math.asinh',
+		atanh   => 'Math.atanh',
+		arctanh => 'Math.arctanh',
+		min     => 'Math.min',
+		max     => 'Math.max',
+		random  => 'Math.random',
+		e       => 'Math.E',
+		pi      => 'Math.PI',
+		'^'     => '**',
+		%vars
+	);
+	my $out = '';
+	my $match;
+
+	$func =~ s/\s//g;
+	while (length($func) > 0) {
+		if (($match) = ($func =~ m/^([A-Za-z]+|\^)/)) {
+			$func = substr($func, length($match));
+			if ($tokens{$match}) {
+				$out .= $tokens{$match};
+			} else {
+				Value::Error("Unknown token $match in function.");
+			}
+		} elsif (($match) = ($func =~ m/^([^A-Za-z^]+)/)) {
+			$func = substr($func, length($match));
+			$out .= $match;
+		} else {    # Shouldn't happen, but to stop an infinite loop for safety.
+			Value::Error("Unknown error parsing function.");
+		}
+	}
+	return "return $out;";
+}
+
+# JavaScript Functions Output
+sub genJS {
+	my $self = shift;
+	return '' unless ($self->{funcType} =~ /^js/);
+	my $id    = shift || 1;
+	my $count = shift || 1;
+	my $vars  = join(', ', @{ $self->{variables} });
+	my $JSout = <<END_OUTPUT;
+	var xData${id}_$count = [];
+	var yData${id}_$count = [];
+	var zData${id}_$count = [];
```

Comment:
There is no `var` in modern javascript.

Reply:
Thanks, I still get confused on when I can use `const` vs `let` when I want to be able to modify the variables. Since I'm just pushing to the array, `const` seems to work. I was just following the plotly docs, many places have not updated things correctly.

Reply:
Use `const` anytime you never assign to the variable again.  Use `let` if you need to assign to it again after initialization.  That is the only criteria.